pytorch_lightning/core/lightning.py

```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
             optimizer: Current optimizer used in training_loop
             optimizer_idx: Current optimizer idx in training_loop
         """
+
+        # Iterate over all optimizer parameters to preserve their `requires_grad` information
+        # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        # make sure current optimizer is latest to be iterated over.
-        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
-        num_optimizers = len(optimizers) - 1
-        for opt_idx, opt in enumerate(optimizers):
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
             for group in opt.param_groups:
                 for param in group['params']:
-                    if num_optimizers == opt_idx:
-                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
-                        if param in param_requires_grad_state:
-                            param.requires_grad = param_requires_grad_state[param]
+                    # If a param appears in multiple optimizers, use the `requires_grad` info already set
+                    if param in param_requires_grad_state:
+                        param.requires_grad = param_requires_grad_state[param]
```

Comment:
Not sure about this line, shouldn't just to be `continue` ? 
With 3 optimizers,
If param is shared only with 1 and 2 optimizers, and current is 3, it is possible you will reset param to True when it should have been False.
Can you check this ?